docs/xedit.1
.\" generated with Ronn/v0.7.3
.\" http://github.com/rtomayko/ronn/tree/0.7.3
.
.TH "XEDIT" "1" "May 2017" "" ""
.
.SH "NAME"
\fBxedit\fR \- a text editor inspired by acme(1) from Plan 9 and Inferno
.
.SH "SYNOPSIS"
\fBxedit\fR [\fIfile\fR\.\.\.]
.
.SH "DESCRIPTION"
\fBxedit\fR is a text editor inspired by the Acme editor from the Plan 9 and Inferno operating systems\. Unlike Acme, \fBxedit\fR is a single window, single file editor\. Instead of baking a window manager into the editor, this job is relegated to an X11 window manager\. It is recommended that \fBxedit\fR be used with a tiling window manager such as dwm(1) or spectrwm(1)\. These window managers will make dealing with multiple windows much easier and more efficient\.
.
.SS "Windows"
\fBxedit\fR windows are divided into four basic regions: a one\-line status, an expanding tags region, a main content region and an adjacent scroll region\. The status region contains a set of symbolic flags indicating the current state of the editor followed by the path of the file being edited\. The tags region acts as scratch buffer for commands that can be executed to affect the file or the state of the editor\. As the content of this region grows it will expand up to a quarter of the size of the window shrinking the main coantent region in kind\. The main content region displays a view of the file currently being edited\. To the left of the content region is a narrow vertical region matching the height of the content region\. This region acts as a scrollbar for the content region\.
.
.SS "Scrolling"
The scroll bar is located to the left of the content region\. It\'s operation is similar to that of acme(1) or 9term(1)\. A left click on the scroll bar will scroll the adjacent line in the content region to the bottom of the window\. A right click will move the adjacent line in the content region to the top of the window\. A middle click will jump to an approximate location in the file determined by calculating the vertical distance from the top of the scrollbar and applying that as a percentage to the offset in the file\.
.
.SS "Typing and Editing"
Typed characters in \fBxedit\fR are delivered to the currently active region\. Which region is active is determined by the placement of the mouse or by keyboard shortcut\. That is to say, the focus follows the mouse much as it does in acme(1) but there is a keyboard shortcut that allows users to toggle focus between the content region and the tag region without using the mouse\.
.
.P
The mechanics of editing text in the tag and content region is identical with the exception of searching and saving\. Edited content in the tag region is not saved to disk unlike the content region\. This region is considered a scratch buffer for commands, notes, and other bits of text that are placed there temporarily\. The content region displays the current view of the file being edited and can be flushed to disk when requested\.
.
.P
Searching with a term selected in the tags region will search for the term in the content region rather than the tags region\. In this way a user can edit the search term incrementally and perform repeated searches through the content region\. Searching for a term in the content region \fIwill\fR search for the term in the content region however\.
.
.SH "TEXT SELECTION"
\fBxedit\fR uses a series of rules to determine how much text to select when the user executes a context sensitive selection, a search, or a context sensitive execution\. The following rules are applied in order until a match is found\.
.
.IP "1." 4
\fBCursor over \'(\' or \')\'\fR: Highlight text between the parentheses including nested parentheses\.
.
.IP "2." 4
\fBCursor over \'[\' or \']\'\fR: Highlight text between the brackets including nested brackets\.
.
.IP "3." 4
\fBCursor over \'{\' or \'}\'\fR: Highlight text between the braces including nested braces\.
.
.IP "4." 4
\fBCursor at beginning or end of line\fR: Highlight the entire line (including the newline)
.
.IP "5." 4
\fBCursor over alphanumeric character or underscore\fR: Highlight the word under the cursor consisting of only alphanumeric and underscore characters\.
.
.IP "" 0
.
.P
If none of the above rules match, \fBxedit\fR will simply highlight the block of non\-whitespace characters under the cursor\.
.
.SH "MOUSE HANDLING"
.
.TP
\fBLeft Button\fR
The left mouse button is used for selecting text or moving the cursor\. A single\-click will move the mose to the clicked location\. A double\-click will will select the object under the clicked location based on context as described in \fBTEXT SELECTION\fR\. A triple\-click will select the largest contiguous chunk of non\-whitespace characters at the clicked location\.
.
.TP
\fBMiddle Button\fR
The middle mouse button is used for executing text at the clicked location\. The command to be executed is determined by the context rules defined in the \fBTEXT SELECTION\fR section\. The cursor position is not changed on a middle click\.
.
.TP
\fBRight Button\fR
The right button is used to search for the next occurrence of the clicked text\. The search term is determined by the context rules defined in the \fBTEXT SELECTION\fR section\. The search direction follows the direction of the previous search operation\. The \fBShift\fR key can be held in combination with a click of the right mosue button in order to reverse the search direction\.
.
.SH "COMMAND EXECUTION"
\fBxedit\fR allows for the execution of any arbitrary text as a command\. The input and output to/from each command executed can be controlled by prepending one of a set of sigils defined below\. These sigils instruct \fBxedit\fR from where the command will receive its input and where it will place its output (both standard and errors)\.
.
.TP
\fB\'!\' \- Run command detached from editor\fR
The command will be executed in the background and all of its input and output file descriptors will be closed\.
.
.TP
\fB\'<\' \- Input from command\fR
The command will be executed in the background and its standard output will be placed in the content region\. Its error output will be placed in the tags region\.
.
.TP
\fB\'>\' \- Output to command\fR
The command will be executed in the background\. The currently selected text will be written to the command\'s standard input\. The command\'s standard output and standard error content will be written to the tags region\.
.
.TP
\fB\'|\' \- Pipe through command\fR
The command will be executed in the background\. The currently selected text will be written to the command\'s standard input\. The command\'s standard output will replace the currently selected text\. Any error output will be placed in the tags region\.
.
.TP
\fB\':\' \- Pipe through sed(1)\fR
Identical to \'|\' except that the command is always sed(1)\. This is a convenience shortcut to allow quick and easy access to sed for editing blocks of text\.
.
.TP
\fBCommands with no sigil\fR
Commands with none of the aforementioned sigils will be executed in the background and have their standard output placed in the content region and their error output placed in the tags region\.
.
.SH "KEYBOARD SHORTCUTS"
.
.SS "Unix Standard Shortcuts"
.
.TP
\fBCtrl+u\fR
Delete from the cursor position to the beginning of the line\.
.
.TP
\fBCtrl+k\fR
Delete from the cursor position to the end of the line\.
.
.TP
\fBCtrl+w\fR
Delete the word to the left\.
.
.TP
\fBCtrl+a\fR
Move cursor to the beginning of the line\.
.
.TP
\fBCtrl+e\fR
Move cursor to the end of the line\.
.
.SS "Cursor Movement and Selection"
The key combinations below are responsible for moving the cursor around the document by character, by word, or by line\. The \fBShift\fR modifier key can be applied to any of them to also extend the current selection to the new cursor position\.
.
.TP
\fBEscape\fR
Highlight the last contiguous block of inserted text or clear the current selection (deselect the currently selected text)\.
.
.TP
\fBLeft\fR
Move the cursor one character to the left\.
.
.TP
\fBRight\fR
Move the cursor one character to the right\.
.
.TP
\fBUp\fR
Move the cursor to the previous line\.
.
.TP
\fBDown\fR
Move the cursor to the next line\.
.
.TP
\fBCtrl+Up\fR
Move the current line or selection up a line\.
.
.TP
\fBCtrl+Down\fR
Move the current line or selection down a line\.
.
.TP
\fBCtrl+Left\fR
Move the cursor to the beginning of the word to the left\.
.
.TP
\fBCtrl+Right\fR
Move the cursor to the end of the word to the right\.
.
.SS "Modern Text Editing Shortcuts"
.
.TP
\fBCtrl+s\fR
Save the contents of the content region to disk\.
.
.TP
\fBCtrl+z\fR
Undo the last change performed on the active region\.
.
.TP
\fBCtrl+y\fR
Redo the previously undone change on the active region\.
.
.TP
\fBCtrl+x\fR
Cut the selected text to the X11 CLIPBOARD selection\. If no text is selected then the current line is cut\.
.
.TP
\fBCtrl+c\fR
Copy the selected text to the X11 CLIPBOARD selection\. If no text is selected then the current line is copied\.
.
.TP
\fBCtrl+v\fR
Paste the contents of the X11 CLIPBOARD selection to the active region\.
.
.TP
\fBCtrl+j\fR
Join the current line and the next line\.
.
.TP
\fBCtrl+l\fR
Select the current line\.
.
.TP
\fBDelete\fR
Delete the character to the right\.
.
.TP
\fBCtrl+Delete\fR
Delete the word to the right\.
.
.TP
\fBBackspace\fR
Delete the character to the left\.
.
.TP
\fBCtrl+Backspace\fR
Delete the word to the left\.
.
.TP
\fBCtrl+Enter\fR
Create a new line after the current line and place the cursor there\.
.
.TP
\fBCtrl+Shift+Enter\fR
Create a new line before the current line and place the cursor there\.
.
.TP
\fBPageUp\fR
Scroll the active region up by one screenful of text\. The cursor is not affected by this operation\.
.
.TP
\fBPageDn\fR
Scroll the active region down by one screenful of text\. The cursor is not affected by this operation\.
.
.SS "Bookmark Shortcuts"
\fBxedit\fR supports marking locations in a document to quickly jump to later\. This eases navigation between multiple locations in a large document\.
.
.TP
\fBCtrl+[0\-9]\fR
Jump to a bookmarked location\.
.
.TP
\fBCtrl+Alt+[0\-9]\fR
Save the cursor location as a bookmark
.
.SS "Search Shortcuts"
The shortcuts below allow the user to search for selected text or by context\. The direction of the search defaults to the forward direction with regard to the position in the file\. Each search follows the direction of the previous search unless the \fBShift\fR modifier is applied\. The \fBShift\fR modifier causes the current search operation to be applied in the opposite direction of the previous\.
.
.TP
\fBCtrl+f\fR
Search for the next occurrence of the selected text in the content region\. If no text is currently selected, the text under the cursor is selected based on context as described in \fBTEXT SELECTION\fR\.
.
.TP
\fBCtrl+Alt+f\fR
Search for the next occurence previous search term in the content region\.
.
.SS "Implementation\-specific Shortcuts"
.
.TP
\fBCtrl+[\fR
Decrease the indent level of the selected text\.
.
.TP
\fBCtrl+]\fR
Increase the indent level of the selected text\.
.
.TP
\fBCtrl+h\fR
Highlight the item under cursor following the rules in \fBTEXT SELECTION\fR
.
.TP
\fBCtrl+t\fR
Toggle focus between the tags region and the content region\.
.
.TP
\fBCtrl+q\fR
Quit the editor\. If the file is modified a warning will be printed in the tags region and the editor will not quit\. Executing the shortcut twice within 250ms will ignore the warning and quit the editor without saving\.
.
.TP
\fBCtrl+d\fR
Execute the selected text as described in \fBCOMMAND EXECUTION\fR\. If no text is selected, the text under cursor is selecte dbased on context as described in \fBTEXT SELECTION\fR\.
.
.TP
\fBCtrl+o\fR
Launch xfilepick(1) to choose a file from a recursive list of files in the current deirectory and sub directories\. This file will be opened in a new instance of \fBxedit\fR\.
.
.TP
\fBCtrl+p\fR
Launch xtagpick(1) to select a tag from a ctags(1) generated index file\. \fBxedit\fR will jump to the selected ctag definition in the current window if the file is currently being edited\. Otherwise, a new instance of \fBxedit\fR will be launched with the target file and the cursor set to the line containing the definition\.
.
.TP
\fBCtrl+g\fR
Lookup the selected symbol or symbol under the cursor in a ctags(1) generated index file\. Jump to the location of the definition if it exist in the current file\. Otherwise, a new instance of \fBxedit\fR will be launched with the target file and the cursor set to the line containing the definition\.
.
.TP
\fBCtrl+Shift+g\fR
Jump to the previous cursor location\.
.
.TP
\fBCtrl+n\fR
Open a new instance of \fBxedit\fR with no filename\.
.
.SH "BUILTINS"
.
.TP
\fBCut\fR
Cut the selection to the X11 CLIPBOARD selection\.
.
.TP
\fBCopy\fR
Copy the selection to the X11 CLIPBOARD selection\.
.
.TP
\fBEol\fR
Toggle the line\-ending style for the buffers contents between LF and CRLF
.
.TP
\fBFind [term]\fR
Find the next occurrence of the selected text\.
.
.TP
\fBGoTo [arg]\fR
Jump to a specific line number or symbol\.
.
.TP
\fBIndent\fR
Toggle the autoindent feature on or off\.
.
.TP
\fBOverwrite\fR
Save the file to disk even if the file has been modified externally\.
.
.TP
\fBPaste\fR
Paste the contents of the X11 CLIPBOARD selection into the buffer\.
.
.TP
\fBQuit\fR
Quit the editor\.
.
.TP
\fBRedo\fR
Redo the last undone change\.
.
.TP
\fBReload\fR
Reload the buffer contents from the on\-disk file\.
.
.TP
\fBSave\fR
Save the contents of the buffer to disk\.
.
.TP
\fBSaveAs [path]\fR
Save the contents of the buffer to disk\. If a path argument is provided, the buffer will be saved to the new path\.
.
.TP
\fBTabs\fR
Toggle the expand tabs featuer on or off\.
.
.TP
\fBUndo\fR
Undo the previous edit\.
.
.SH "FILES"
.
.TP
\fB$HOME/\.config/edit/editrc\fR
Shell script loaded in current environment to make shell functions and environment variables available to xedit(1)
.
.SH "ENVIRONMENT"
.
.TP
\fBEDITTAGS\fR
The contents of this environment variable is used to initialize the contents of the tags region of the window\.
.
.TP
\fBSHELL\fR
The contents of this variable are used as the shell in which all non\-builtin commands are executed\. If this variable is not defined, sh(1) is used as a fallback shell\.
.
.SH "AUTHOR"
Michael D\. Lowis
.
.SH "SEE ALSO"
xedit(1) xpick(1) xfilepick(1) xtagpick(1)


xedit.sync-conflict-20170524-185737.c
#include <stdc.h>
#include <x11.h>
#include <utf.h>
#include <edit.h>
#include <ctype.h>
#include <win.h>
#include <shortcuts.h>

typedef struct {
    char* tag;
    union {
        void (*noarg)(void);
        void (*arg)(char* arg);
    } action;
} Tag;

/* The shell: Filled in with $SHELL. Used to execute commands */
static char* ShellCmd[] = { NULL, "-c", NULL, NULL };
static char* SedCmd[] = { "sed", "-e", NULL, NULL };
static char* PickFileCmd[] = { "xfilepick", ".", NULL };
static char* PickTagCmd[] = { "xtagpick", NULL, "tags", NULL, NULL };
static char* OpenCmd[] = { "xedit", NULL, NULL };
static Tag Builtins[];
static int SearchDir = DOWN;
static char* SearchTerm = NULL;

/* Tag/Cmd Execution
 ******************************************************************************/
static Tag* tag_lookup(char* cmd) {
    size_t len = 0;
    Tag* tags = Builtins;
    for (char* tag = cmd; *tag && !isspace(*tag); tag++, len++);
    while (tags->tag) {
        if (!strncmp(tags->tag, cmd, len))
            return tags;
        tags++;
    }
    return NULL;
}

static void tag_exec(Tag* tag, char* arg) {
    /* if we didnt get an arg, find one in the selection */
    if (!arg) arg = view_getstr(win_view(TAGS), NULL);
    if (!arg) arg = view_getstr(win_view(EDIT), NULL);
    /* execute the tag handler */
    tag->action.arg(arg);
    free(arg);
}

static void cmd_exec(char* cmd) {
    char op = '\0';
    if (*cmd == ':' || *cmd == '!' || *cmd == '<' || *cmd == '|' || *cmd == '>')
        op = *cmd, cmd++;
    ShellCmd[2] = cmd;
    /* execute the command */
    char *input = NULL, *output = NULL, *error = NULL;
    WinRegion dest = EDIT;
    if (op && op != '<' && op != '!' && 0 == view_selsize(win_view(EDIT)))
        win_view(EDIT)->selection = (Sel){ .beg = 0, .end = buf_end(win_buf(EDIT)) };
    input = view_getstr(win_view(EDIT), NULL);

    if (op == '!') {
        cmdrun(ShellCmd, NULL);
    } else if (op == '>') {
        dest = TAGS;
        output = cmdwriteread(ShellCmd, input, &error);
    } else if (op == '|') {
        output = cmdwriteread(ShellCmd, input, &error);
    } else if (op == ':') {
        SedCmd[2] = cmd;
        output = cmdwriteread(SedCmd, input, &error);
    } else {
        if (op != '<') dest = win_getregion();
        output = cmdread(ShellCmd, &error);
    }

    if (error)
        view_append(win_view(TAGS), chomp(error));

    if (output) {
        if (op == '>')
            view_append(win_view(dest), chomp(output));
        else
            view_putstr(win_view(dest), output);
        win_setregion(dest);
    }
    /* cleanup */
    free(input);
    free(output);
    free(error);
}

static void exec(char* cmd) {
    /* skip leading space */
    for (; *cmd && isspace(*cmd); cmd++);
    if (!*cmd) return;
    /* see if it matches a builtin tag */
    Tag* tag = tag_lookup(cmd);
    if (tag) {
        while (*cmd && !isspace(*cmd++));
        tag_exec(tag, (*cmd ? stringdup(cmd) : NULL));
    } else {
        cmd_exec(cmd);
    }
}

/* Action Callbacks
 ******************************************************************************/
static void onerror(char* msg) {
    view_append(win_view(TAGS), msg);
}

static void trim_whitespace(void) {
    Buf* buf = win_buf(EDIT);
    if (TrimOnSave && buf_end(buf) > 0) {
        View* view = win_view(EDIT);
        unsigned off = 0, prev = 1;
        /* loop through the buffer till we hit the end or we stop advancing */
        while (off < buf_end(buf) && prev != off) {
            off = buf_eol(buf, off);
            Rune r = buf_get(buf, off-1);
            for (; (r == ' ' || r == '\t'); r = buf_get(buf, off-1)) {
                if (off <= view->selection.beg) {
                    view->selection.end--;
                    view->selection.beg--;
                }
                off = buf_delete(buf, off-1, off);
            }
            /* make sure we keep advancing */
            prev = off;
            off  = buf_byline(buf, off, +1);
        }
    }
}

static void quit(void) {
    static uint64_t before = 0;
    uint64_t now = getmillis();
    if (!win_buf(EDIT)->modified || (now-before) <= 250) {
        #ifndef TEST
        x11_deinit();
        #else
        exit(0);
        #endif
    } else {
        view_append(win_view(TAGS),
            "File is modified. Repeat action twice in < 250ms to quit.");
    }
    before = now;
}

static bool changed_externally(Buf* buf) {
    bool modified = (buf->modtime != modtime(buf->path));
    if (modified) {
        view_append(win_view(TAGS),
            "File modified externally: Reload, Overwrite, or {SaveAs }");
    }
    return modified;
}

static void overwrite(void) {
    trim_whitespace();
    buf_save(win_buf(EDIT));
}

static void save(void) {
    if (!changed_externally(win_buf(EDIT)))
        overwrite();
}

static void reload(void) {
    view_reload(win_view(EDIT));
}

/* Mouse Handling
 ******************************************************************************/
void onmouseleft(WinRegion id, bool pressed, size_t row, size_t col) {
    static int count = 0;
    static uint64_t before = 0;
    if (!pressed) return;
    uint64_t now = getmillis();
    count = ((now-before) <= 250 ? count+1 : 1);
    before = now;

    if (count == 1) {
        if (x11_keymodsset(ModShift))
            view_selext(win_view(id), row, col);
        else
            view_setcursor(win_view(id), row, col);
    } else if (count == 2) {
        view_select(win_view(id), row, col);
    } else if (count == 3) {
        view_selword(win_view(id), row, col);
    }
}

void onmousemiddle(WinRegion id, bool pressed, size_t row, size_t col) {
    if (pressed) return;
    if (win_btnpressed(MouseLeft)) {
        cut();
    } else {
        char* str = view_fetch(win_view(id), row, col);
        if (str) exec(str);
        free(str);
    }
}

void onmouseright(WinRegion id, bool pressed, size_t row, size_t col) {
    if (pressed) return;
    if (win_btnpressed(MouseLeft)) {
        paste();
    } else {
        SearchDir *= (x11_keymodsset(ModShift) ? -1 : +1);
        free(SearchTerm);
        SearchTerm = view_fetch(win_view(id), row, col);
        if (view_findstr(win_view(EDIT), SearchDir, SearchTerm)) {
            win_setregion(EDIT);
            win_warpptr(EDIT);
        }
    }
}

/* Keyboard Handling
 ******************************************************************************/
static void saveas(char* arg) {
    if (arg) {
        char* path = win_buf(EDIT)->path;
        win_buf(EDIT)->path = stringdup(arg);
        buf_save(win_buf(EDIT));
        free(path);
    }
}

static void tag_undo(void) {
    view_undo(win_view(EDIT));
}

static void tag_redo(void) {
    view_redo(win_view(EDIT));
}

static void search(void) {
    char* str;
    SearchDir *= (x11_keymodsset(ModShift) ? -1 : +1);
    if (x11_keymodsset(ModAlt) && SearchTerm)
        str = stringdup(SearchTerm);
    else
        str = view_getctx(win_view(FOCUSED));
    view_findstr(win_view(EDIT), SearchDir, str);
    free(SearchTerm);
    SearchTerm = str;
    if (view_selsize(win_view(EDIT))) {
        win_setregion(EDIT);
        win_warpptr(EDIT);
    }
}

static void execute(void) {
    char* str = view_getcmd(win_view(FOCUSED));
    if (str) exec(str);
    free(str);
}

static void find(char* arg) {
    SearchDir *= (x11_keymodsset(ModShift) ? -1 : +1);
    view_findstr(win_view(EDIT), SearchDir, arg);
}

static void open_file(void) {
    char* file = cmdread(PickFileCmd, NULL);
    if (file) {
        file = chomp(file);
        if (!win_buf(EDIT)->path && !win_buf(EDIT)->modified) {
            buf_load(win_buf(EDIT), file);
        } else {
            OpenCmd[1] = file;
            cmdrun(OpenCmd, NULL);
        }
    }
    free(file);
}

static void pick_symbol(char* symbol) {
    PickTagCmd[1] = "fetch";
    PickTagCmd[3] = symbol;
    char* pick = cmdread(PickTagCmd, NULL);
    if (pick) {
        Buf* buf = win_buf(EDIT);
        if (buf->path && 0 == strncmp(buf->path, pick, strlen(buf->path))) {
            view_setln(win_view(EDIT), strtoul(strrchr(pick, ':')+1, NULL, 0));
            win_setregion(EDIT);
        } else {
            if (!buf->path && !buf->modified) {
                view_init(win_view(EDIT), pick, onerror);
            } else {
                OpenCmd[1] = chomp(pick);
                cmdrun(OpenCmd, NULL);
            }
        }
    }
}

static void pick_ctag(void) {
    pick_symbol(NULL);
}

static void complete(void) {
    View* view = win_view(FOCUSED);
    buf_getword(&(view->buffer), risword, &(view->selection));
    view->selection.end = buf_byrune(&(view->buffer), view->selection.end, RIGHT);
    PickTagCmd[1] = "print";
    PickTagCmd[3] = view_getstr(view, NULL);
    char* pick = cmdread(PickTagCmd, NULL);
    if (pick)
        view_putstr(view, chomp(pick));
    free(PickTagCmd[3]);
}

static void jump_to(char* arg) {
    if (arg) {
        size_t line = strtoul(arg, NULL, 0);
        if (line) {
            view_setln(win_view(EDIT), line);
            win_setregion(EDIT);
        } else {
            pick_symbol(arg);
        }
    }
}

static void goto_ctag(void) {
    char* str = view_getctx(win_view(FOCUSED));
    jump_to(str);
    free(str);
}

static void tabs(void) {
    bool enabled = !(win_buf(EDIT)->expand_tabs);
    win_buf(EDIT)->expand_tabs = enabled;
    win_buf(TAGS)->expand_tabs = enabled;
}

static void indent(void) {
    bool enabled = !(win_buf(EDIT)->copy_indent);
    win_buf(EDIT)->copy_indent = enabled;
    win_buf(TAGS)->copy_indent = enabled;
}

static void del_indent(void) {
    view_indent(win_view(FOCUSED), LEFT);
}

static void add_indent(void) {
    view_indent(win_view(FOCUSED), RIGHT);
}

static void eol_mode(void) {
    int crlf = win_buf(EDIT)->crlf;
    win_buf(EDIT)->crlf = !crlf;
    win_buf(TAGS)->crlf = !crlf;
    exec(crlf ? "|dos2unix" : "|unix2dos");
}

static void new_win(void) {
    cmd_exec("!edit");
}

static void newline(void) {
    View* view = win_view(FOCUSED);
    if (x11_keymodsset(ModShift)) {
        view_byline(view, UP, false);
        view_bol(view, false);
        if (view->selection.end == 0) {
            view_insert(view, true, '\n');
            view->selection = (Sel){0,0,0};
            return;
        }
    }
    view_eol(view, false);
    view_insert(view, true, '\n');
}

void highlight(void) {
    view_selctx(win_view(FOCUSED));
}

/* Main Routine
 ******************************************************************************/
static Tag Builtins[] = {
    { .tag = "Cut",       .action.noarg = cut       },
    { .tag = "Copy",      .action.noarg = copy      },
    { .tag = "Eol",       .action.noarg = eol_mode  },
    { .tag = "Find",      .action.arg   = find      },
    { .tag = "GoTo",      .action.arg   = jump_to   },
    { .tag = "Indent",    .action.noarg = indent    },
    { .tag = "Overwrite", .action.noarg = overwrite },
    { .tag = "Paste",     .action.noarg = paste     },
    { .tag = "Quit",      .action.noarg = quit      },
    { .tag = "Redo",      .action.noarg = tag_redo  },
    { .tag = "Reload",    .action.noarg = reload    },
    { .tag = "Save",      .action.noarg = save      },
    { .tag = "SaveAs",    .action.arg   = saveas    },
    { .tag = "Tabs",      .action.noarg = tabs      },
    { .tag = "Undo",      .action.noarg = tag_undo  },
    { .tag = NULL,        .action.noarg = NULL      }
};

static KeyBinding Bindings[] = {
    /* Cursor Movements */
    { ModAny, KEY_HOME,  cursor_home  },
    { ModAny, KEY_END,   cursor_end   },
    { ModAny, KEY_UP,    cursor_up    },
    { ModAny, KEY_DOWN,  cursor_dn    },
    { ModAny, KEY_LEFT,  cursor_left  },
    { ModAny, KEY_RIGHT, cursor_right },

    /* Standard Unix Shortcuts */
    { ModCtrl, 'u', del_to_bol  },
    { ModCtrl, 'k', del_to_eol  },
    { ModCtrl, 'w', del_to_bow  },
    { ModCtrl, 'a', cursor_bol  },
    { ModCtrl, 'e', cursor_eol  },

    /* Standard Text Editing Shortcuts */
    { ModCtrl, 's', save        },
    { ModCtrl, 'z', undo        },
    { ModCtrl, 'y', redo        },
    { ModCtrl, 'x', cut         },
    { ModCtrl, 'c', copy        },
    { ModCtrl, 'v', paste       },
    { ModCtrl, 'j', join_lines  },
    { ModCtrl, 'l', select_line },

    /* Block Indent */
    { ModCtrl, '[', del_indent },
    { ModCtrl, ']', add_indent },

    /* Common Special Keys */
    { ModNone, KEY_PGUP,      page_up   },
    { ModNone, KEY_PGDN,      page_dn   },
    { ModAny,  KEY_DELETE,    delete    },
    { ModAny,  KEY_BACKSPACE, backspace },

    /* Implementation Specific */
    { ModNone,                 KEY_ESCAPE, select_prev  },
    { ModCtrl,                 't',        change_focus },
    { ModCtrl,                 'q',        quit         },
    { ModCtrl,                 'h',        highlight    },
    { ModCtrl,                 'f',        search       },
    { ModCtrl|ModShift,        'f',        search       },
    { ModCtrl|ModAlt,          'f',        search       },
    { ModCtrl|ModAlt|ModShift, 'f',        search       },
    { ModCtrl,                 'd',        execute      },
    { ModCtrl,                 'o',        open_file    },
    { ModCtrl,                 'p',        pick_ctag    },
    { ModCtrl,                 'g',        goto_ctag    },
    { ModCtrl,                 'n',        new_win      },
    { ModCtrl,                 '\n',       newline      },
    { ModCtrl|ModShift,        '\n',       newline      },
    { ModCtrl,                 ' ',        complete     },
    { 0, 0, 0 }
};

void onscroll(double percent) {
    size_t bend = buf_end(win_buf(EDIT));
    size_t off  = (size_t)((double)bend * percent);
    view_scrollto(win_view(EDIT), (off >= bend ? bend : off));
}

void onfocus(bool focused) {
    /* notify the user if the file has changed externally */
    (void)changed_externally(win_buf(EDIT));
}

void onupdate(void) {
    static char status_bytes[256];
    memset(status_bytes, 0, sizeof(status_bytes));
    char* status = status_bytes;
    Buf* buf = win_buf(EDIT);
    *(status++) = (buf->charset == BINARY ? 'B' : 'U');
    *(status++) = (buf->crlf ? 'C' : 'N');
    *(status++) = (buf->expand_tabs ? 'S' : 'T');
    *(status++) = (buf->copy_indent ? 'I' : 'i');
    *(status++) = (SearchDir < 0 ? '<' : '>');
    *(status++) = (buf->modified ? '*' : ' ');
    *(status++) = ' ';
    char* path = (buf->path ? buf->path : "*scratch*");
    size_t remlen = sizeof(status_bytes) - strlen(status_bytes) - 1;
    strncat(status, path, remlen);
    win_settext(STATUS, status_bytes);
    win_view(STATUS)->selection = (Sel){0,0,0};
}

void onlayout(void) {
    /* calculate and update scroll region */
    View* view = win_view(EDIT);
    size_t bend = buf_end(win_buf(EDIT));
    if (bend == 0) bend = 1;
    if (!view->rows) return;
    size_t vbeg = view->rows[0]->off;
    size_t vend = view->rows[view->nrows-1]->off + view->rows[view->nrows-1]->rlen;
    double scroll_vis = (double)(vend - vbeg) / (double)bend;
    double scroll_off = ((double)vbeg / (double)bend);
    win_setscroll(scroll_off, scroll_vis);
}

void onshutdown(void) {
    quit();
}

#ifndef TEST
int main(int argc, char** argv) {
    /* setup the shell */
    ShellCmd[0] = getenv("SHELL");
    if (!ShellCmd[0]) ShellCmd[0] = "/bin/sh";
    /* Create the window and enter the event loop */
    win_window("edit", onerror);
    char* tags = getenv("EDITTAGS");
    win_settext(TAGS, (tags ? tags : DEFAULT_TAGS));
    win_setruler(80);
    view_init(win_view(EDIT), (argc > 1 ? argv[1] : NULL), onerror);
    win_setkeys(Bindings);
    win_loop();
    return 0;
}
#endif
